### DISCUSSION TITLE
E3onCAN bricht mit Fehler ab

### DISCUSSION THREAD

**Discussion by neandr** [MyHomeMyData/E3onCAN#12] [Q&A]
Beim Versuch E3onCAN mit folgendem Befehl zu starten, gibt's nur 'n Fehler:

```
~/E3onCAN $ python3 E3onCANcollect.py -c can0 -dev vx3 -r 1690,1834 -m 192.xxx.xxx.xxx:1883:open3e
Traceback (most recent call last):
  File "/home/pi/E3onCAN/E3onCANcollect.py", line 263, in <module>
    client_mqtt = paho.Client("E3onCANclient.py")
  File "/home/pi/.local/lib/python3.9/site-packages/paho/mqtt/client.py", line 772, in __init__
    raise ValueError(
ValueError: Unsupported callback API version: version 2.0 added a callback_api_version, see docs/migrations.rst for details
```

In meiner Installation auf einem Raspberry steht im File
```
File "/home/pi/E3onCAN/E3onCANcollect.py", line 263, in <module>
    client_mqtt = paho.Client("E3onCANclient.py")
```
Dieses Programm finde ich nirgends!? E3onCAN wurde per ZIP von der GIT Seite in ein neues Verzeichnis `E3onCAN` geladen.
Ausser `pip3 install -r requirements.txt` finde ich keine weitere Installationsanweisung.

Der RPI sollte auf neuestem Stand sein nach `sudo apt-get update`  und `sudo apt-get upgrade`

Wie kann E3onCANcollect erfolgreich eingesetzt werden?

**Comment by MyHomeMyData** [comment]
Welche Version von paho-mqtt ist installiert?

```
pip show paho-mqtt
```

**Reply by neandr** [reply]
```
~ $ pip show paho-mqtt

Name: paho-mqtt
Version: 2.1.0
Summary: MQTT version 5.0/3.1.1 client class
Home-page: None
Author: None
Author-email: Roger Light <roger@atchoo.org>
License: EPL-2.0 OR BSD-3-Clause
Location: /home/pi/.local/lib/python3.9/site-packages
Requires: 
Required-by: open3e

```

**Reply by MyHomeMyData** [reply]
Hm, das sieht ok aus. Würdest Du bitte mal E3onCAN in einer virtuellen Umgebung testen? Das geht mit
```
sudo apt install python3-virtualenv
mkdir ~/E3onCANvenv && cd ~/E3onCANvenv
python3 -m venv .venv && source .venv/bin/activate
git clone https://github.com/MyHomeMyData/E3onCAN.git
cd E3onCAN
pip3 install -r requirements.txt
```
Dann bitte nochmal den ursprünglichen Aufruf von E3onCANcollect ausprobieren.

Falls Du kein Git installiert hast, zu allererst noch ein `sudo apt install git` machen.

**Comment by neandr** [comment]
> Dann bitte nochmal den ursprünglichen Aufruf von E3onCANcollect ausprobieren.

Ja das läuft länglich durch.
Aber ich kann auf der Page keinen  Hinweis entdecken, dass das "vorher" ausgeführt werden muss/sollte

**Reply by MyHomeMyData** [reply]
Gut, dann gibt es in Deiner Python-Umgebung vermutlich einen Konflikt mit pip-Bibliotheken, die direkt installiert wurden. Das lässt sich zuverlässig vermeiden, wenn man mit venv arbeitet.

Das muss man aber nicht tun. E3onCAN sollte auch funktionieren, wenn man's direkt installiert und keine Bibliotheks-Konflikte entstehen oder bereits vorhanden sind.

Aber danke für den Hinweis. Ich werde eine entsprechende Empfehlung im Readme ergänzen.

**Comment by neandr** [comment]
Aus dem Listing an der Konsole hatte ich den Eindruck, dass sich die Abfragewerte wiederholen, im wesentlichen wohl E380.
Habe das mal abgebrochen, und nochmal den eigentlichen Befehl abgesetzt.
Leider immer noch der beschriebene Fehler. hmmm

**Reply by MyHomeMyData** [reply]
Jetzt bin ich verwirrt. Es funktioniert auch nicht in der virtuellen Umgebung?

> Aus dem Listing an der Konsole hatte ich den Eindruck, dass sich die Abfragewerte wiederholen

E3onCANcollect macht keinerlei Abfragen. Es dekodiert nur CAN-Messages, die es auf dem Bus mitliest. Der E380 sendet halt sehr viele Messages, ohne dass man fragen müsste.

**Comment by neandr** [comment]
Sehe gerade es soll in einer Virt.Umgebung getest werden. Das hatte ich übersehen. Mach's jetzt mal

**Comment by neandr** [comment]
siehe oben 
`pip3 install -r requirements`
sollte wohl sein:
`pip3 install -r requirements.txt`

**Comment by neandr** [comment]
OK, nach der Korrektur des letzten Inst.Befehls läuft dann auch 
```
~/E3onCANvenv/E3onCAN $ python3 E3onCANcollect.py -c can0 -dev vx3 -m 192.168.178.70:1883:open3e -mfstr {device}_{didNumber:04d}_{didName} -retain 1834
Wait for dids and publish to mqtt...
```
und in Node-Red werden die Daten -- ohne weitere Aufforderung -- eingelesen, also "Collect" Betrieb.  Dies aber "nur" für VX3. 

Wie an anderer Stelle besprochen, bin ich ja auf der Suche, wie mit allen Geräten auf dem CAN-Bus gelesen und dann auch geschrieben werden kann.

**Reply by MyHomeMyData** [reply]
> Dies aber "nur" für VX3.

Wenn Du -dev vx3 angiebst, werden nur die Daten der VX3 dekodiert. Ohne Angabe von -dev wird alles dekodiert, was dekodierbar ist.

Es ist auch möglich, E3onCANcollect mehrfach zu starten und jeweils mit -dev auf eine anderes Device zu filtern.

Schreiben geht mit open3e und ioBroker.e3oncan. Aber das haben wir ja schon besprochen.

**Reply by MyHomeMyData** [reply]
Ich fürchte, das mit -dev weglassen stimmt nicht. Die Angabe von -dev ist immer notwendig. Sorry, habe ich verwechselt.

**Comment by neandr** [comment]
Das mit dem <ins>ohne</ins> `-dev` ist gut, aber mir fehlt `vcal` 
Auch bei der Angabe `-dev vcal` kommt nix. Habe mit Vicare einige Werte für die Wärmepumpe geändert, auch "Einmalige Warmwasserbereitung". Obwohl die Temperatur steigt, es werden aber keine Änderungswerte per MQTT übertragen.
Gibt es eine Monitormöglichkeit was intern abgeht?

**Reply by MyHomeMyData** [reply]
Die Vitocal sendet von sich aus nur auf dem internen CAN-Bus. Da müsstest Du umklemmen. VX3 funktioniert da auch, E380 aber nicht.

**Comment by neandr** [comment]
Das bedeutet, dass vcal Werte gepollt werden wollen?
Oder wirklich 2.CAN-Bus Adapter an den internen CAN-Bus der WP?

**Reply by MyHomeMyData** [reply]
Wenn Du "Collect" auf der vcal und dem E380 nutzen möchtest, ja, 2 Adapter. Für VX3 und vcal reicht es, auf den internen Bus umzuklemmen.

Oder auf dem externen bleiben und für vcal open3e nutzen. [Hier](https://github.com/MyHomeMyData/ioBroker.e3oncan/discussions/34) hab ich versucht, das mal zu erläutern.
